src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonAdvancedConverter.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json.Serialization.Metadata;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal abstract class JsonAdvancedConverter<T> : JsonResumableConverter<T>
+    {
+        internal virtual bool IsJsonDictionaryConverter => false;
+
+        internal virtual bool OnWriteResume(Utf8JsonWriter writer, T dictionary, JsonSerializerOptions options, ref WriteStack state)
+        {
+            Debug.Fail("We should have overriden it if we got here");
+            return true;
+        }
+    }
+
+    /// <summary>
+    /// Base class for object, enumerable and dictionary converters
+    /// </summary>
+    internal abstract class JsonAdvancedConverter<T, IntermediateType> : JsonAdvancedConverter<T>
```

Comment:
Nit: would avoid terms like "advanced" in type names since they don't necessarily convey what this layer of abstraction is meant to achieve.

Reply:
didn't have a better name, I'm open for feedback